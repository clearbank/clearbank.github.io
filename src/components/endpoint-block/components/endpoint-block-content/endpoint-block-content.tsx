import React, { ReactElement } from "react"

import { Collapse, CollapseItem } from "src/components/clearbank-ui"
import * as DataRow from "src/components/data-row/data-row.styles"

import * as Types from "./endpoint-block-content.types"

const isRequired = (key: string, requiredItems: string[]) =>
  requiredItems.some((requiredItem) => requiredItem === key)

/**
 * Get group labels for display on-screen
 * @param propertyName The property to be displayed
 * @param groups The available groups defined from oneOf
 * @returns The string to display that defines the groups
 */
const getOneOfGroup = (propertyName: string, groups: object) => {
  const appearsIn = groups[propertyName]

  if (!appearsIn) {
    return ""
  }

  return "Group: " + appearsIn.map((group: number) => group + 1).join(", ")
}

const EndpointBlockContent: React.FunctionComponent<
  Types.EndpointBlockContentProps
> = ({
  allProperties,
  requiredPropertyKeys,
  propertyGroups = {},
}): ReactElement => {
    // whitelisted attributes
    const additionalAttributes = [
      "minimum",
      "maximum",
      "minLength",
      "maxLength",
      "pattern",
      "enum",
    ]
    const attributeNameMap = {
      maximum: "Maximum",
      minimum: "Minimum",
      minLength: "Minimum length",
      maxLength: "Maximum length",
      pattern: "Pattern",
      enum: "Enum array",
    }

    return (
      <>
        {Object.keys(allProperties).map((propertyKey: string) => {
          const currentProperty = allProperties[propertyKey]

          let { type, description, items, properties, required, allOf, oneOf } =
            allProperties[propertyKey]

          const isOneOf = oneOf && Array.isArray(oneOf) && oneOf.length > 0

          // Merge properties and required into one object
          if (isOneOf) {
            required = oneOf.reduce(
              (output: Array<Object>, choice: Types.OneOfProps) =>
                output.concat(choice.required || []),
              []
            )
            properties = oneOf.reduce(
              (output: Array<Object>, choice: Types.OneOfProps) =>
                Object.assign(output, choice.properties),
              {}
            )
          }

          const hasNestedProperties = !!properties
          const hasNestedItems = !!items && !!items.properties
          const hasAdditionalAttributes = additionalAttributes.some(
            (additionalAttribute) => additionalAttribute in currentProperty
          )
          let showAdditionalAttributes = false
          // Check if schema to combine - schema can have multiple models
          // Note, for more complex scenarios a more intelligent algorithm is required
          const hasAnyOf = allOf && allOf.length > 0

          // Group properties by oneOf groups and create of a map of property to group
          // The same name could appear in both groups so ensure there is an array

          // This will be used to then seed the display with information on mutually
          // exclusive groups. Note this doesn't currently deal with recursion i.e. a oneOf
          // nested within a oneOf - that needs some thought
          const oneOfPropertyGroups: object = (isOneOf ? oneOf : []).reduce(
            (output: object, choice: Types.OneOfProps, index: number) => {
              if (choice.properties) {
                Object.keys(choice.properties).forEach((prop) => {
                  output[prop] = (output[prop] || []).concat([index])
                })
              }

              return output
            },
            {}
          )

          let parentObjectGuidance = isOneOf
            ? oneOf.length < 2
              ? ""
              : oneOf.reduce(
                (output: string, choice: Types.OneOfProps, index: number) => {
                  const choicePropertyNames = Object.keys(choice.properties || [])

                  if (choicePropertyNames.length === 0) {
                    return output + " the value of the property with the type " + choice.type
                  }

                  const requirementDescription = choicePropertyNames
                    .map((propertyName, index) => {
                      if (index !== 0 && index === choicePropertyNames.length - 1) {
                        return ", and " + propertyName
                      }

                      if (index !== 0 && index === choicePropertyNames.length - 2) {
                        return ", " + propertyName
                      }

                      return propertyName
                    })

                  if (index === 0) {
                    return (output + requirementDescription.join(""))
                  }

                  if (index !== oneOf.length - 1) {
                    return output + ", " + requirementDescription.join("")
                  }

                  return (output + " OR " + requirementDescription.join("") + " as shown in the groups below.")
                },
                " You need to include EITHER "
              )
            : ""

          let allOfSchema: any = {}

          if (hasAnyOf) {
            description = allOf[0].description // description is always from first key
            // remaining schema models are merged, i.e enums. Only supports one schema
            allOfSchema = allOf.reduce(
              (result: Array<string>, current: Object) => {
                return Object.assign(result, current)
              },
              {}
            )
            type = allOfSchema.type // Layout only supports single type
          }

          if (hasAdditionalAttributes || hasAnyOf || isOneOf) {
            showAdditionalAttributes = true
          }

          return (
            <DataRow.DataListItem key={propertyKey}>
              <DataRow.Header className="great-primer">
                {propertyKey}
              </DataRow.Header>
              {type && (
                <>
                  <DataRow.Type> {type}</DataRow.Type>
                </>
              )}
              {requiredPropertyKeys &&
                isRequired(propertyKey, requiredPropertyKeys) && (
                  <>
                    {type && <>,{"  "}</>}
                    <DataRow.Required>Required</DataRow.Required>
                  </>
                )}
              {/* new line */}
              {
                <>
                  <DataRow.Description>
                    {getOneOfGroup(propertyKey, propertyGroups)}
                  </DataRow.Description>
                </>
              }
              {description && (
                <>
                  <DataRow.Description>
                    {description + parentObjectGuidance}
                  </DataRow.Description>
                </>
              )}
              {showAdditionalAttributes &&
                additionalAttributes.map((additionalAttribute) => {
                  const targetAttributes = hasAnyOf
                    ? allOfSchema
                    : currentProperty
                  const hasAttributeFromList =
                    additionalAttribute in targetAttributes

                  if (!hasAttributeFromList) {
                    return null
                  }

                  const attributeKey =
                    additionalAttribute in attributeNameMap
                      ? attributeNameMap[additionalAttribute]
                      : additionalAttribute // fallback
                  const attributeValue = targetAttributes[additionalAttribute]
                  const isComplexType = Array.isArray(attributeValue)

                  return (
                    <DataRow.AdditionalAttributesList key={additionalAttribute}>
                      <DataRow.AdditionalAttributeKey>
                        {attributeKey}
                      </DataRow.AdditionalAttributeKey>
                      <DataRow.AdditionalAttributeValue
                        isComplexType={isComplexType}
                      >
                        {isComplexType
                          ? attributeValue.join(", ")
                          : attributeValue}
                      </DataRow.AdditionalAttributeValue>
                    </DataRow.AdditionalAttributesList>
                  )
                })}

              {(hasNestedProperties || hasNestedItems) && (
                <DataRow.DataList nested>
                  <Collapse>
                    <CollapseItem title="Show children" nested>
                      {hasNestedProperties && (
                        <EndpointBlockContent
                          allProperties={properties}
                          requiredPropertyKeys={required}
                          propertyGroups={oneOfPropertyGroups}
                        />
                      )}
                      {hasNestedItems && (
                        <EndpointBlockContent
                          allProperties={items.properties}
                          requiredPropertyKeys={items.required}
                          propertyGroups={oneOfPropertyGroups}
                        />
                      )}
                    </CollapseItem>
                  </Collapse>
                </DataRow.DataList>
              )}
            </DataRow.DataListItem>
          )
        })}
      </>
    )
  }

export default EndpointBlockContent
