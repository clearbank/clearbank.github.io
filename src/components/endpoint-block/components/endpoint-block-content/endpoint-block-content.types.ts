export interface EndpointBlockContentProps {
  allProperties: any
  requiredPropertyKeys: any
  propertyGroups?: object
}

export interface OneOfProps {
  type: string,
  required: Array<string>,
  properties: Object,
}
