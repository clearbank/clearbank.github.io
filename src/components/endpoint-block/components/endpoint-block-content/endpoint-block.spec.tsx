import React from 'react'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom'
import 'jest-styled-components'

import Component from './endpoint-block-content'

import * as Types from './endpoint-block-content.types'

let component: ReturnType<typeof render> = null
let fragment: DocumentFragment = null
let root: HTMLElement = null

const allProperties: any = {
  "status": {
    "description": "The status of the account.",
    "enum": ["NotProvided", "Enabled", "Closed", "Suspended"],
    "type": "string"
  },
  "statusReason": {
    "description": "The account status reason",
    "enum": ["NotProvided", "AccountHolderBankrupt", "AccountHolderDeceased", "AccountSwitched", "CompanyNoLongerTrading", "DissatisfiedCustomer", "DuplicateSoleTraderAccount", "FinancialCrime", "FraudFirstParty", "FraudThirdParty", "Other"],
    "type": "string"
  },
  "sourceAccount": {
    "oneOf": [
      {
        "type": "string"
      },
      {
        "type": "object",
        "required": ["iban"],
        "properties": {
          "iban": {
            "type": "string",
            "minLength": 1,
            "pattern": "[A-Z]{2,2}[0-9]{2,2}[a-zA-Z0-9]{1,30}",
          }
        },
        "additionalProperties": false
      },
      {
        "type": "object",
        "required": ["schemeName", "identification"],
        "properties": {
          "schemeName": {
            "type": "string",
            "enum": ["BBan", "Other"],
            "example": "Other"
          },
          "proprietary": {
            "type": "string",
            "enum": ["SortcodeAccountNumber"],
          },
          "identification": {
            "type": "string",
            "minLength": 1,
            "example": "01020301234567"
          }
        },
        "additionalProperties": false
      }
    ]
  }
}

const defaultProps: Types.EndpointBlockContentProps = {
  allProperties,
  requiredPropertyKeys: ['status']
}

beforeEach(() => {
  component = render(<Component {...defaultProps} children='children' />)

  fragment = component.asFragment()
  root = component.container.firstChild
})

afterEach(() => {
  fragment = null
  root = null
})

describe('Component', () => {
  test('matches snapshot with enum array and status as a required field', () => {
    expect(root).toMatchSnapshot()
  })
})
