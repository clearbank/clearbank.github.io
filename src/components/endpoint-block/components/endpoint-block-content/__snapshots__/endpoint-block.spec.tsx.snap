// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component matches snapshot with enum array and status as a required field 1`] = `
.c0 {
  padding: 15px 0 15px 20px;
  border-top: dashed 1px #E0E0E0;
}

.c2 {
  font-size: 13px;
  color: #178880;
  font-family: clear-sans-text,"Roboto Mono",monospace;
}

.c1 {
  font-size: 15px;
  font-weight: bold;
  font-family: clear-sans-text,"Roboto Mono",monospace;
  text-transform: capitalize;
  color: #000000;
}

.c3 {
  font-size: 13px;
  text-transform: uppercase;
  color: #D00202;
  font-family: clear-sans-text,"Roboto Mono",monospace;
}

.c4 {
  margin-bottom: 0;
  font-size: 16px;
  word-wrap: break-word;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  line-height: 1.5;
  font-size: 13px;
  font-family: clear-sans-text,"Roboto Mono",monospace;
  color: #000000;
}

.c6 {
  -webkit-flex: 0 1 auto;
  -ms-flex: 0 1 auto;
  flex: 0 1 auto;
  white-space: nowrap;
}

.c6::after {
  content: ':';
  padding-right: 5px;
}

.c7 {
  -webkit-flex: 1 0 0;
  -ms-flex: 1 0 0;
  flex: 1 0 0;
}

.c7::before {
  content: '[';
}

.c7::after {
  content: ']';
}

<li
  class="c0"
>
  <span
    class="c1 great-primer"
  >
    status
  </span>
  <span
    class="c2"
  >
     
    string
  </span>
  ,
    
  <span
    class="c3"
  >
    Required
  </span>
  <p
    class="c4"
  />
  <p
    class="c4"
  >
    The status of the account.
  </p>
  <dl
    class="c5"
  >
    <dt
      class="c6"
    >
      Enum array
    </dt>
    <dd
      class="c7"
    >
      NotProvided, Enabled, Closed, Suspended
    </dd>
  </dl>
</li>
`;
